torch/distributed/_shard/checkpoint/filesystem.py

```diff
@@ -0,0 +1,127 @@
+        super().__init__()
+        self.path = path
+        os.makedirs(self.path, exist_ok=True)
+
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
+                storage.write(req.bytes.getbuffer())
+
+        fut: Future[None] = Future()
+        fut.set_result(None)
+        return fut
+
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
+        for req in requests:
+            with open(os.path.join(self.path, req.storage_key), "wb") as storage:
+                # The following couple lines are simple implementation to get
+                # things going.
+                #
+                # At load time, to enable resharding, we use (sub)view of the tensor.
+                # Since the storage of the tensor might not be contiguous. we need to
+                # preseve the original view, to calculate the correct sub view at load.
+                #
+                # `torch.save` saves both the view and storage, it is a good option
+                # for unblocking. There are two drawbacks:
+                # 1. `torch.save` is pickle based, and pickle is not known for its
+                #   compatibility, we should consider replacing it with a more
+                #   stable option.
+                # 2. pickle is not streamable.
+                buffer = io.BytesIO()
+                torch.save(req.tensor, buffer)
+                storage.write(buffer.getbuffer())
+
+        fut: Future[None] = Future()
+        fut.set_result(None)
+        return fut
+
+    # Implementating the abstract function in Storage Writer
+    def write_metadata(self, metadata: Metadata) -> None:
+        # Once write metadata once as each Metadata has the global view
+        if dist.is_initialized() and dist.get_rank() != 0:
+            return
+
+        with open(os.path.join(self.path, ".metadata"), "wb") as metadata_file:
+            pickle.dump(metadata, metadata_file)
+
+class FileSystemReader(StorageReader):
+    def __init__(self, path: str) -> None:
+        super().__init__()
+        self.path = path
+
+    def read_tensors(self, requests: List[TensorReadRequest]) -> Future[None]:
+        """
+        Very basic implementation that read from file system.
+        """
+        # Sort the the requests by storage key and try to reuse the loaded tensors
+        requests.sort(key=operator.attrgetter("storage_key"))
+
+        cached_storage_key = None
+        view_cached: Optional[Tensor] = None
+
+        for req in requests:
+            if cached_storage_key != req.storage_key or \
+                    (view_cached is not None and view_cached.device != req.tensor.device):
+
+                with open(os.path.join(self.path, req.storage_key), "rb") as storage:
+                    view_cached = cast(Tensor, torch.load(storage, map_location=req.tensor.device))
```

Comment:
We only produce Tensor reads when the checkpoint has a tensor related entry.
The ways this could fail are either due to bugs in serialization or an adversarial checkpoint.